src/lmdb.ml

```diff
@@ -298,6 +334,28 @@ module Make (Key : Values.S) (Elt : Values.S) = struct
   type key = Key.t
   type elt = Elt.t
 
+  let dbval_of_buffer b =
+    let mvp = addr (make mdb_val) in
+    alive_while b mvp; (* Make sure buffer will stay alive while mvp is in use. *)
+    (mvp |-> mv_size) <-@ Bigarray.Array1.dim b ;
+    (mvp |-> mv_data) <-@ to_voidp @@ bigarray_start array1 b ; (* Reference to b is lost here! *)
+    (*Gc.full_major () ;*) (* trigger possible use-after-free. *)
+    mvp
+  let buffer_of_dbval mvp =
+    (* no need to keep a reference to mvp here, because the memory the bigarray
+     * is mapped to is in the lmdb map and only valid as long as the transaction
+     * is alive. The user knows this. *)
+    bigarray_of_ptr array1
+      (!@ (mvp |-> mv_size))
+      Char
+      (!@ (mvp |-> mv_data) |> from_voidp char)
+
+  let write f v =
+    let _alloc len =
+      Bigarray.(Array1.create char C_layout len)
+    in
+    f v
```

Comment:
It seems to me `dbval_of_buffer` and `write` are always used together. You might as well call `dbval_of_buffer` inside `write` directly.

Reply:
This won't be the case anymore when using ``MDB_RESERVE``.

Reply:
So much foreshadowing and teasing! ;)